Sources/SwiftDocC/Model/Rendering/RenderSectionTranslator/DeclarationsSectionTranslator.swift

```diff
@@ -114,3 +208,224 @@ struct DeclarationsSectionTranslator: RenderSectionTranslator {
         }
     }
 }
+
+fileprivate extension DeclarationRenderSection.Token {
+    /// Whether this token is a text token with a changed highlight.
+    var isHighlightedText: Bool {
+        self.highlight == .changed && self.kind == .text
+    }
+
+    /// Whether this token has any highlight applied to it.
+    var isHighlighted: Bool {
+        self.highlight != nil
+    }
+
+    /// Create a new ``Kind/text`` token with the given text and highlight.
+    init(plainText text: String, highlight: Highlight? = nil) {
+        self.init(text: text, kind: .text, highlight: highlight)
+    }
+}
+
+/// "Post-process" a sequence of declaration tokens by recombining text tokens and cleaning up
+/// highlighted spans of tokens.
+///
+/// The output of `preProcessFragment` is suitable for differencing, but causes unexpected results
+/// when handed directly to Swift-DocC-Render and applied to its declaration formatter. This
+/// function recombines adjacent tokens so that the declaration formatter can correctly see the
+/// structure of a declaration and format it accordingly.
+///
+/// In addition, this function takes the opportunity to trim whitespace at the beginning and end of
+/// highlighted spans of tokens, to beautify the resulting declaration.
+fileprivate func postProcessTokens(
+    _ tokens: [DeclarationRenderSection.Token]
+) -> [DeclarationRenderSection.Token] {
+    var processedTokens: [DeclarationRenderSection.Token] = []
+
+    // This function iterates a list of tokens, keeping track of a "previous token" and comparing it
+    // against the "current token" to perform the following transformations:
+    //
+    // 1. Trim whitespace from the end of a highlighted token if the following token is not
+    //    highlighted.
+    // 2. Trim whitespace from the beginning of a highlighted token if the previous token was not
+    //    highlighted.
+    // 3. Concatenate plain-text tokens if they are both highlighted or both not highlighted.
+
+    /// If the given tokens are both plain-text tokens with the same highlight, return a token with
+    /// both tokens' texts.
+    func concatenateTokens(
+        _ lhs: DeclarationRenderSection.Token,
+        _ rhs: DeclarationRenderSection.Token
+    ) -> DeclarationRenderSection.Token? {
+        guard lhs.kind == .text, rhs.kind == .text, lhs.highlight == rhs.highlight else {
+            return nil
+        }
+        var result = lhs
+        result.text += rhs.text
+        return result
+    }
+
+    var previousToken: DeclarationRenderSection.Token? = nil
+    for var currentToken in tokens {
```

Comment:
Minor: if you write this as 
```suggestion
    guard var previousToken = tokens.first else { return [] }
    for var currentToken in tokens.dropFirst() {
```
then you can reduce the indentation of the entire body by getting rid of `if var previousToken = previousToken`